Add EAD collection import

solth opened the pull request:
This pull request adds the option to import EAD collections from uploaded XML files. The import creates one (parent) process for the collection and one (child) for _each_ "file" contained in the XML file, thus allowing complete collections ("Bestände) into individual processes from a single file upload.

To be able to use this EAD import, first a ruleset and an Import configuration of type "File Upload" has to be configured with metadata format `EAD`:
<img width="1502" alt="Bildschirmfoto 2024-11-08 um 00 34 29" src="https://github.com/user-attachments/assets/42e07a56-5147-497d-a071-f4b30efc691b">

Accordingly, the mapping file used in the import configuration has to map EAD to the Kitodo internal metadata format:
<img width="1492" alt="Bildschirmfoto 2024-11-08 um 00 39 34" src="https://github.com/user-attachments/assets/ccaccaf6-ff7d-42b2-9def-c89e351d8f3c">

A separate mapping file can be configured to map the collection level of the imported EAD XML file, in the field "Mapping file for parent record":
<img width="1490" alt="Bildschirmfoto 2024-11-08 um 00 41 17" src="https://github.com/user-attachments/assets/8449ba7d-5d46-4605-a1de-5434a71f8cd2">

This pull request adds example mapping files for the parent and child processes (`eadParent2kitodo.xsl` and `ead2kitodo.xsl`) as well as a corresponding ruleset (`ruleset_ead.xml`) that can be used for this purpose and that of cause can also extended when required.

The user can then use this import configuration to upload an EAD XML file containing the EAD collection. The upload dialog also allows to select the EAD level of the element that is to be transformed into the parent process from `COLLECTION`, `CLASS` and `SERIES` and the EAD level of items that are transformed into child processes from `FILE` and `ITEM`:
<img width="569" alt="Bildschirmfoto 2024-11-08 um 00 30 33" src="https://github.com/user-attachments/assets/508d78bb-02c4-45eb-8693-50d2a2d15c71">

Additionally, a limit for the maximum number or child processes that can be processed interactively in the import mask can now be configured in `kitodo_config.properties` using the parameter `maxNumberOfProcessesForImportMask`. (this value defaults to 5). If the number of child elements in the uploaded file exceeds this limit, the user is offered with a choice to relocate the import of all processes to a background task:
<img width="698" alt="Bildschirmfoto 2024-11-08 um 00 51 50" src="https://github.com/user-attachments/assets/efbd6f5e-5fff-4de1-bd30-183a34b17923">

If the user has the corresponding permission, the progress of this background import can then be monitored on the "Task manager" page of Kitodo:
<img width="1497" alt="Bildschirmfoto 2024-11-08 um 00 53 15" src="https://github.com/user-attachments/assets/eded309d-98f4-4b3c-b91d-886d76dffe4f">


Fixes #5984

henning-gerhardt commented on Kitodo/src/main/java/org/kitodo/production/forms/createprocess/CreateProcessForm.java:
```diff
@@ -108,6 +122,10 @@ public CreateProcessForm() {
         this.priorityList = priorityList;
     }
 
+    public void calculateNumberOfEadElements() throws XMLStreamException {
```
Missing JavaDoc

henning-gerhardt commented on Kitodo/src/main/java/org/kitodo/production/forms/createprocess/CatalogImportDialog.java:
```diff
@@ -360,4 +397,59 @@ public void setAdditionalImport(boolean additionalImport) {
+    /**
+     * Set selectedField.
+     *
+     * @param field as java.lang.String
```
see above.

henning-gerhardt commented on Kitodo/src/main/java/org/kitodo/production/model/LazyHitModel.java:
```diff
@@ -19,25 +19,21 @@
      * Empty default constructor. Sets default catalog and search field, if configured.
      */
-    public LazyHitModel() {
+    public LazyHitModel(CatalogImportDialog dialog) {
```
Javadoc should be adjusted.

henning-gerhardt commented on Kitodo/src/main/java/org/kitodo/production/services/data/ImportService.java:
```diff
@@ -589,21 +613,46 @@ private void importParents(String recordId, ImportConfiguration importConfigurat
-        if (Objects.isNull(parentID)) {
-            this.parentTempProcess = null;
-            return;
+    public void checkForParent(String parentID, Ruleset ruleset, int projectID) {
```
JavaDoc should be adjusted.

henning-gerhardt commented on Kitodo/src/main/java/org/kitodo/production/thread/ImportEadProcessesThread.java:
```diff
@@ -0,0 +1,348 @@
+    }
+
+    private void cleanUpProcesses(List<Integer> processIds, int newParentId) {
+        logger.error("Deleting processes created until this point to resolve erroneous intermediate state.");
```
Is `error` log level correct? Should not log level `info` or lower be used here? It did not look like that an error occur and get logged.

henning-gerhardt commented on Kitodo/src/main/java/org/kitodo/production/forms/createprocess/CatalogImportDialog.java:
```diff
@@ -360,4 +397,59 @@ public void setAdditionalImport(boolean additionalImport) {
+    /**
+     * Set searchTerm.
+     *
+     * @param searchTerm as java.lang.String
```
`java.lang.String` could be simplified to ` string` or the whole wording should be adjusted.

henning-gerhardt commented on Kitodo/src/main/java/org/kitodo/production/services/data/ImportService.java:
```diff
@@ -679,19 +728,116 @@ public LinkedList<TempProcess> getChildProcesses(ImportConfiguration importConfi
+                StringConstants.LEVEL, level);
+    }
+
+    public boolean isMaxNumberOfRecordsExceeded(String xmlString, String tagName) throws XMLStreamException {
```
Missing JavaDoc.

henning-gerhardt commented on Kitodo/src/main/java/org/kitodo/production/services/data/ImportService.java:
```diff
@@ -409,8 +412,7 @@ public String getParentID(Document document, String higherLevelIdentifier, Strin
                                                      int templateID, int projectID)
-            throws ProcessGenerationException, IOException, TransformerException, InvalidMetadataValueException,
-            NoSuchMetadataFieldException {
+            throws ProcessGenerationException, IOException, TransformerException {
```
JavaDoc should be adjusted.

henning-gerhardt reviewed (commented):
Some notes from reading the changes.

BartChris commented:
While testing this i first made a dumb configuration error. I associated the wrong XSL-file with the EAD-Mapping-File.  
This brought up the following exception message:

![KITODO_EAD_ERROR](https://github.com/user-attachments/assets/3c3c55ca-5145-4ce0-b4ac-2b96c0b78c75)

This error is produced by the fact that the `internalRecord` does only contain the XML header and no real content at that point. 

https://github.com/kitodo/kitodo-production/blob/c947be25295c3e7a3cc09ef7764b8037b32b55b9/Kitodo/src/main/java/org/kitodo/production/services/data/ImportService.java#L894

It is a dumb mistake on my side, but maybe we can assist by also check if we also check if the importDocument is empty and the importer has to check the mapping file configuration.

BartChris commented:
I tested with the given test ead file. One problem is, when i try to import the file two times. I then get a Null pointer for the  `createProcessLocation` call of the child processes and they are not created.
The problem is that when `processTempProcess` is called for a child process

https://github.com/kitodo/kitodo-production/blob/c947be25295c3e7a3cc09ef7764b8037b32b55b9/Kitodo/src/main/java/org/kitodo/production/services/data/ImportService.java#L1227

the process titles are considered invalid because they do already exist. 
https://github.com/kitodo/kitodo-production/blob/c947be25295c3e7a3cc09ef7764b8037b32b55b9/Kitodo/src/main/java/org/kitodo/production/services/data/ImportService.java#L1342-L1344

This should throw a `ProcessGenerationException`, but it seems this exeption does not stop the execution  in the CreateProcessForm. The execution continues and we have processes without an ID. 

The the Nullpointer is thrown here, because the file location is based on the missing ID.

https://github.com/kitodo/kitodo-production/blob/c947be25295c3e7a3cc09ef7764b8037b32b55b9/Kitodo/src/main/java/org/kitodo/production/services/file/FileService.java#L242-L246

When processing the list in the background (because the number of imported processes is greater than the configured limit) the process stops with a ProcessGenerationException.

The behaviour should probably be more like in the mass import, where items with uplicate titles are skipped.

BartChris commented on Kitodo/src/main/java/org/kitodo/production/services/data/ImportService.java:
```diff
@@ -679,19 +736,125 @@ public LinkedList<TempProcess> getChildProcesses(ImportConfiguration importConfi
+
+        if (parentElements.size() != 1) {
+            throw new ProcessGenerationException(
+                    String.format("EAD XML does not contain exactly one element of parent level '%s'!",
```
This should be translated.

solth replied:
Done.

BartChris commented:
While inspecting the Pull request i noticed a general problem:
https://github.com/kitodo/kitodo-production/issues/6303

solth commented on Kitodo/src/main/java/org/kitodo/production/helper/XMLUtils.java:
```diff
@@ -179,4 +195,97 @@ public static void validateXPathSyntax(String xpathString) throws XPathExpressio
+     * @return number of EAD elements with given level
+     * @throws XMLStreamException when parsing XML string fails
+     */
+    public static int getNumberOfElements(String xmlString, String eadLevel) throws XMLStreamException {
```
@BartChris yes, you are right. This was intended to be used in more general imports, but at the moment only works for the EAD import. I have renamed the method as you suggested.

solth commented:
> When processing the list in the background (because the number of imported processes is greater than the configured limit) the process stops with a ProcessGenerationException.
> 
> For the background processing the behaviour should probably be more like in the mass import, where items with duplicate titles are skipped. In the form Kitodo should catch invalid titles and force the user to correct them.

@BartChris thanks for your analysis. Actually I implemented it this way deliberately. The goal here was to restore the original state - from _before_ the import started - when an error occurs so that the user can first check the uploaded file for problems, fix those problems and then re-try the upload with the updated file. Therefore, all processes that were already created up to the point where the exception is thrown are "cleaned up", e.g. removed.

I do understand, though, that a differnt behavior could be preferable, where importing collections does not fail with individual problematic EAD data records, but instead skips those processes and proceeds with the next "items" from the EAD file. If there is no consense on which option is better, we could ideally introduce a parameter in `kitodo_config.properties` to control the error handling that way.

BartChris commented:
> I tested with the given test ead file. A problem appears, when i try to import the file two times. I then get a Null pointer for the `createProcessLocation` call of the child processes and they are not created. The problem is that when `processTempProcess` is called for a child process (with an already existing title)
> 

I still have the problem, that when using the standard import way (without background task) i get a NullPointer exception, when i try to import the same EAD collection multiple times.

solth commented:
> > I tested with the given test ead file. A problem appears, when i try to import the file two times. I then get a Null pointer for the `createProcessLocation` call of the child processes and they are not created. The problem is that when `processTempProcess` is called for a child process (with an already existing title)
> 
> I still have the problem, that when using the standard import way (without background task) i get a NullPointer exception, when i try to import the same EAD collection multiple times.

This seems to be a problem in the hierarchical import in general that also occurs in the master branch and is not related to the changes in this pull request. There (in the master branch), I just tried to import a process hierarchy from Kalliope (ID `DE-2060-BE-13-C35`, import depth = 1, import children = true) _twice_ and on the second try got the same `NullPointerException` you documented above. 

@BartChris If this is easy to resolve I can try to include a fix in this pull request, otherwise I would suggest to handle that issue separately.

solth commented:
> I still have the problem, that when using the standard import way (without background task) i get a NullPointer exception, when i try to import the same EAD collection multiple times.

@BartChris I think this will take a little more time to properly fix. I would therefor suggest to deal with this problem separately. I will create a ticket for this problem.

BartChris commented:
@solth I agree and approve the Pull request. I think the real usage in the Archives have to prove, wether we need additional adjustments, but it works good for me.

 Some additional ideas for the future:

- Also allow the background processing for the "normal" mass import
- Think about a way to provide a report after a background processed mass import (EAD and normal) to report the result of the processing  (Skipped processes? Reason?) in a way which is usable for non-admins

solth commented:
> Some additional ideas for the future:
> 
>     * Also allow the background processing for the "normal" mass import
> 
>     * Think about a way to provide a report after a background processed mass import (EAD and normal) to report the result of the processing  (Skipped processes? Reason?) in a way which is usable for non-admins

Yes, these are definitely two incremental steps that we should try to achieve and that I already had in mind as well. Thanks a lot for the review!
